runtime/doc/treesitter.txt

```diff
@@ -1479,6 +1482,21 @@ anyway as they will be very frequent. Rather a plugin that does any kind of
 analysis on a tree should use a timer to throttle too frequent updates.
 
 
+LanguageTree:async_parse({range}, {opts})         *LanguageTree:async_parse()*
+    Like |LanguageTree:parse()|, but asynchronous.
+
+    Parameters: ~
+      • {range}  (`boolean|Range?`) Parse this range in the parser's source.
+                 Set to `true` to run a complete parse of the source (Note:
+                 Can be slow!) Set to `false|nil` to only parse regions with
+                 empty ranges (typically only the root tree without
+                 injections).
+      • {opts}   (`AsyncParseOpts?`) Options:
```

Comment:
Seems natural for this to be on the existing `:parse()`. If a callback isn't provided then it's synchronous.
```suggestion
      • {opts}   (`ParseOpts?`) Options:
      * {on_parse} (fun(trees: table<integer, TSTree>)?)
```

Reply:
Yes, that's what I originally did.

Reply:
Yeah I am fine to change it back but I did make it two separate functions for a few reasons:

1. Combining them would have to change the function signature of `parse()` from `Trees` to `Trees?`. This is not that intuitive because a sync parse always returns a list of trees and an async parse always returns `nil`. Things are slightly more confusing now, and returning the `Trees` in the async parse (to prevent function signature change) is out of the question because the trees are not yet parsed and it would be bad to return this to a user that might otherwise think async parsing was completed
2. There are times where one might just want to do something like `:async_parse()` (no arguments). If we combine the functions we will have to pass an empty callback `function() end`. Maybe `{ async = true }` would be acceptable actually...
3. If #19624 is ever completed and we want to change the async version to return a promise (or equivalent), it might be preferable to have two different functions (each with different coloring) so we can know exactly when it is safe to, say, `:await()` the result and when we can just use our returned `trees` immediately. Same issue as 1. I guess

Reply:
Most of the UV functions combine sync and async into the same signature and it's generally been a well received design. Changing the static signature to return `tree?` isn't a big deal. If no callback is given then it still always returns `tree`. I also don't think passing an empty callback is a big deal either.

Reply:
Ok that sounds perfect, thanks a ton :+1: